src/util/cluster.js

```diff
@@ -1,122 +1,136 @@
+ * - `CLUSTER_MANAGER_MAX_STUCK_WORKER_RESPAWNS`: Maximum number of times stuck workers will be respawned (default: 3)
+ * - `STATS_CLIENT`: If set to 'prometheus', a metrics app will be started on a different port METRICS_PORT (default: 9091).
+ *
+ * @param {*} port the port to start the app on
+ * @param {*} app the app instance
+ * @param {*} metricsApp the app instance for metrics, it will be started on a different port (METRICS_PORT)
+ */
+async function start(port, app, metricsApp) {
   // Enable diagnostic reporting on SIGUSR2 signal for debugging stuck processes
   process.report.reportOnSignal = true;
 
-  if (cluster.isMaster) {
-    logger.info(`Master (pid: ${process.pid}) has started`);
-
-    // HTTP server for exposing metrics
-    if (process.env.STATS_CLIENT === 'prometheus') {
-      const metricsServer = metricsApp.listen(metricsPort);
-
-      gracefulShutdown(metricsServer, {
-        signals: 'SIGINT SIGTERM SIGSEGV',
-        timeout: 30000, // timeout: 30 secs
-        forceExit: false, // triggers process.exit() at the end of shutdown process
-      });
-    }
-
-    // Fork workers.
-    for (let i = 0; i < numWorkers; i += 1) {
-      cluster.fork({
-        WORKER_ID: `worker-${i + 1}`,
-      });
-    }
-
-    cluster.on('online', (worker) => {
-      logger.info(`Worker (pid: ${worker.process.pid}) is online`);
-      // To provide caching at pod-level
-    });
-
-    let isShuttingDown = false;
-    cluster.on('exit', async (worker) => {
-      if (!isShuttingDown) {
-        logger.error(`Worker (pid: ${worker.process.pid}) died`);
-        logger.error(`Killing other workers to avoid any side effects of the dead worker`);
-        logProcessInfo();
-        isShuttingDown = true;
-        await shutdownWorkers();
+  const numWorkers = parseInt(process.env.NUM_PROCS || '1', 10);
+
+  const metricsPort = parseInt(process.env.METRICS_PORT || '9091', 10);
+
+  // Set the maximum number of ms to wait for graceful shutdown
+  const shutdownTimeout = parseInt(process.env.CLUSTER_MANAGER_SHUTDOWN_TIMEOUT || '30000', 10);
+
+  // Set the maximum number of times that a worker that dies unexpectedly will be restarted.
+  // If the worker dies more than this number of times, the cluster will shutdown.
+  const maxKilledWorkerRestarts = parseInt(
+    process.env.CLUSTER_MANAGER_MAX_KILLED_WORKER_RESTARTS || '3',
+    10,
+  );
+
+  // Set the maximum number of times that stuck workers will be respawned.
+  // Any more than this number of times, the cluster will shutdown.
+  const maxStuckWorkerRespawns = parseInt(
+    process.env.CLUSTER_MANAGER_MAX_STUCK_WORKER_RESPAWNS || '3',
+    10,
+  );
+
+  const usePrometheus = process.env.STATS_CLIENT === 'prometheus';
+
+  let server;
+  const clusterManager = new ClusterManager({
+    numWorkers,
+    pingFrequency: 10000, // 10 seconds
+    pingTimeout: 90000, // 90 seconds
```

Comment:
looks bit high? are we planning to gradually decrease it?

Reply:
with batch yield most of integrations finish earlier so this purely metrics aggregator taking time. can we start with 30 seconds